pkg/gui/files_panel.go

```diff
@@ -379,31 +379,61 @@ func (gui *Gui) catSelectedFile(g *gocui.Gui) (string, error) {
 }
 
 func (gui *Gui) pullFiles(g *gocui.Gui, v *gocui.View) error {
-	gui.createMessagePanel(g, v, "", gui.Tr.SLocalize("PullWait"))
+	if err := gui.createMessagePanel(gui.g, v, "", gui.Tr.SLocalize("PullWait")); err != nil {
+		return err
+	}
 	go func() {
-		if err := gui.GitCommand.Pull(); err != nil {
-			gui.createErrorPanel(g, err.Error())
-		} else {
-			gui.closeConfirmationPrompt(g)
-			gui.refreshCommits(g)
-			gui.refreshStatus(g)
+		unamePassOpend := false
+		err := gui.GitCommand.Pull(func(passOrUname string) string {
+			unamePassOpend = true
+			return gui.waitForPassUname(g, v, passOrUname)
+		})
+		if err != nil {
+			errMessage := err.Error()
+			if errMessage == "exit status 128" {
+				errMessage = gui.Tr.SLocalize("PassUnameWrong")
+			}
+			_ = gui.createErrorPanel(g, errMessage)
+		}
+		if unamePassOpend {
+			_, _ = g.SetViewOnBottom("pushPassUname")
+			_ = g.DeleteView("pushPassUname")
+		}
+		if err == nil {
+			_ = gui.closeConfirmationPrompt(g)
+			_ = gui.refreshCommits(g)
+			_ = gui.refreshStatus(g)
 		}
 		gui.refreshFiles(g)
 	}()
 	return nil
 }
 
-func (gui *Gui) pushWithForceFlag(currentView *gocui.View, force bool) error {
-	if err := gui.createMessagePanel(gui.g, currentView, "", gui.Tr.SLocalize("PushWait")); err != nil {
+func (gui *Gui) pushWithForceFlag(g *gocui.Gui, v *gocui.View, force bool) error {
+	if err := gui.createMessagePanel(g, v, "", gui.Tr.SLocalize("PushWait")); err != nil {
 		return err
 	}
 	go func() {
+		unamePassOpend := false
 		branchName := gui.State.Branches[0].Name
-		if err := gui.GitCommand.Push(branchName, force); err != nil {
-			_ = gui.createErrorPanel(gui.g, err.Error())
-		} else {
-			_ = gui.closeConfirmationPrompt(gui.g)
-			_ = gui.refreshSidePanels(gui.g)
+		err := gui.GitCommand.Push(branchName, force, func(passOrUname string) string {
+			unamePassOpend = true
+			return gui.waitForPassUname(g, v, passOrUname)
+		})
+		if err != nil {
```

Comment:
there is some code duplication here between the fetch/push/pull functions, which I think would be good to factor out. I'm not fully sure how to do that given that they rely on a variable used in a closure, but the three different pieces of code have already begun diverging. For example, here we're calling `refreshSidePanels` at the end but pullFiles we're calling `refreshCommits` and `refreshStatus`